### PR TITLE
Handle Quoted Paths

### PR DESCRIPTION
**Pull request by EliSauder**
Since I was in here and the parsing is updated, I figured #13 probably wouldn't be that bad to add. Here is my go at it :)

### CODE REVIEW

File: impl.go [hunk 1]
```diff
@@ -98,13 +103,22 @@ func stripPaths(in string) string {
 			break
 		}
 		runes = runes[n:]
+		if isQuotedPath && len(runes) > 0 {
```
Comment (josharian) [inline impl.go]: Should we check that the skipped rune is “?

File: impl.go [hunk 2]
```diff
@@ -98,13 +103,22 @@ func stripPaths(in string) string {
...
+			runes = runes[1:]
+		}
 
 		// Copy non-path runes verbatim
 		n = slices.IndexFunc(runes, isPathRune)
 		seg = runes
 		if n >= 0 {
 			seg = seg[:n]
 		}
+		if n > 0 && seg[0] == '"' {
```
Comment (josharian) [inline impl.go]: Please add hasPrefix and hasSuffix helpers instead of checking n and n[0]. (Or maybe even better would be trimPrefix and trimSuffix? On my phone.)

File: impl.go [hunk 3]
```diff
@@ -127,7 +141,8 @@ func lastIndex[S ~[]E, E comparable](s S, v E) int {
 // isPathRune reports whether r can appear in an import path.
 // See https://go.dev/ref/spec#Import_declarations.
 func isPathRune(r rune) bool {
-	return unicode.IsPrint(r) && !strings.ContainsRune(" \uFFFD!\"#$%&'()*,:;<=>?[\\]^`{|}", r)
```
Comment (josharian) [inline impl.go:130]: On my phone…what is this diff doing here?

File: impl_test.go [hunk 4]
```diff
@@ -914,8 +915,13 @@ func TestStripPaths(t *testing.T) {
 	}{
 		{desc: "no path", input: "Iface", want: "Iface"},
 		{desc: "simple path", input: "a/b.T", want: "b.T"},
+		{desc: "simple quoted path", input: "\"a/b\".T", want: "b.T"},
```
Comment (josharian) [inline impl_test.go]: This would be nicer if we used backtics for the string literals

**Review by josharian — CHANGES_REQUESTED** [review]
Thanks. Always a good sign when a change makes other fixes easier…

File: impl.go (hunk 1 above)
Comment (EliSauder) [reply impl.go]: My thinking on not checking it is that a path shouldn't otherwise contain `"`, so if it started with one, to be valid, it should also end with one. 
If we updated the stripPaths function to return an error as well, I think adding in validation would make sense.

File: impl.go (hunk 3 above)
Comment (EliSauder) [reply impl.go:130]: Think that was a formatting change that I forgot to revert (I'll change it back in the next commit)

File: impl_test.go (hunk 4 above)
Comment (EliSauder) [reply impl_test.go]: So: `` `"a/b".T` ``?

File: impl.go (hunk 2 above)
Comment (EliSauder) [reply impl.go]: Added trimPathPrefix and trimPathSuffix in the most recent push

File: impl.go (hunk 1 above)
Comment (EliSauder) [reply impl.go]: @josharian is that logic good or should I add error handling?

Comment (josharian) [reply impl.go]: oops, sorry, this got lost in my inbox.

I think we should add these test cases (or ones like them) to `TestFindInterface`:

```go
		// Quoted path edge cases - unbalanced quotes
		{input: `"a/b/c/pkg.Typ`, wantErr: true},              // missing closing quote
		{input: `a/b/c/pkg".Typ`, wantErr: true},              // missing opening quote
		{input: `"a/b/c/pkg.Typ"`, wantErr: true},             // quote after type name
		{input: `""a/b/c/pkg"".Typ`, wantErr: true},           // double quotes
		{input: `"github.com/josharian/impl/testdata".Interface1`, path: "github.com/josharian/impl/testdata", typ: Type{Name: "Interface1"}},
```

and make sure they pass. I'll look at the rest of the now...

File: impl.go [hunk 5]
```diff
@@ -85,6 +85,11 @@ func stripPaths(in string) string {
 	out := make([]rune, 0, len(runes))
 	for len(runes) > 0 {
 		// Find extent of path-like segment
+		isQuotedPath := false
```
Comment (josharian) [inline impl.go]: On another read through, I think we can delete this entirely, as well as the chunk at line 106 under discussion. I don't think this ever triggers.

File: impl.go [hunk 6]
```diff
@@ -98,13 +103,18 @@ func stripPaths(in string) string {
...
+			runes = runes[1:]
+		}
 
 		// Copy non-path runes verbatim
 		n = slices.IndexFunc(runes, isPathRune)
 		seg = runes
 		if n >= 0 {
 			seg = seg[:n]
 		}
+		seg = trimPathPrefix(seg)
```
Comment (josharian) [inline impl.go]: Thanks. This refactor made the structure a lot clearer. AFAICS, these two lines are the ones that matter, not the other bits above. May I suggest:

```go
// seg contains the path. If it is quoted, remove the quotes.
seg = unquote(seg)
```

and then unquote is:

```go
func unquote(p []rune) []rune {
  if len(p) < 2 || p[0] != '"' || p[len(p)-1] != '"' {
    return p
  }
 return p[1:len(p)-1]
```

**Review by josharian — COMMENTED** [review]
what do you think about these simplifcations? (i haven't confirmed they pass all the tests, including the new ones i suggested! apologies if i'm leading you astray...)

File: impl.go (hunk 1 above)
Comment (EliSauder) [reply impl.go]: So I take it, you do want it to have error handling and properly error out on incorrect quotations?

File: impl.go (hunk 5 above)
Comment (EliSauder) [reply impl.go]: I thought the were load bearing, but after removing them all the tests passed (except for the new ones which fail regardless)

File: impl.go (hunk 6 above)
Comment (EliSauder) [reply impl.go]: This won't (and doesn't) work because of how the for loop behaves. The start and end quote will be apart of `seg` on separate iterations.

Basically if we have: `"github.com/foo".Iface`
1. First iteration
  a. Grab all path characters (none, bc " is first)
  b. Grab all non-path characters (")
  c. Strip quotes (remove only the one)
2. Second iteration
  a. Grab all path characters (github.com/foo)
  b. Grab all non-path characters "
  c. Strip quotes
3. etc

Comment (josharian) [reply impl.go]: Oh right. OK, thanks. A comment in here would go a long ways.

File: impl.go (hunk 5 above)
Comment (josharian) [reply impl.go]: Great.

File: impl.go (hunk 1 above)
Comment (josharian) [reply impl.go]: Y’know, maybe it’s not worth the code. Let’s just get this in.

File: impl.go (hunk 5 above)
Comment (EliSauder) [reply impl.go]: Just removed them in the most recent pushes

File: impl.go (hunk 6 above)
Comment (EliSauder) [reply impl.go]: I did simplify it down to one function though since that way it will correctly only remove one quote for double quoted paths.

File: impl.go (hunk 1 above)
Comment (EliSauder) [reply impl.go]: Lol, I just pushed changes that handle the errors for double quotes and unbalanced. However, I think I've found a bug in findInterfaces. 

stripPaths is behaving correctly (even added some extra tests to confirm). But find interfaces is failing on the below tests. After checking with some print logging, the `baseInput[:dot+1]` doesn't send the interface path to strippaths.
 
image

Comment (EliSauder) [reply impl.go]: Or I guess, less of a bug, and more that findInterfaces does its own path parsing outside of stripPaths

Comment (EliSauder) [reply impl.go]: I think this ties a bit into the fact that impl can't currently properly handle multiple paths

Comment (josharian) [reply impl.go]: Yeah, I was idly wondering whether strip paths should run earlier in findInterfaces. But my bandwidth for staring at this is limited right now. As long as we keep building up good e2e tests we can refactor more later. (Or have an agent do it.)

Comment (EliSauder) [reply impl.go]: Sounds good. I'm going to do some clean up and commenting and then we can merge it roughly as is.

Comment (EliSauder) [reply impl.go]: Documents and cleanup in most recent push.

**Comment by EliSauder** [comment]
@josharian I was looking over the stripPaths test cases, why did you decide to have non-path characters terminate the path segment rather than just have them be invalid?

I think it would make more sense for all of these to expect an error.

```go
		{desc: "equals terminates path", input: "Iface[a/b=c.T]", want: "Iface[b=c.T]"},
		// Exclamation mark is excluded - path ends at !
		{desc: "exclamation terminates path", input: "Iface[a/b!c.T]", want: "Iface[b!c.T]"},
		// Semicolon is excluded - path ends at ;
		{desc: "semicolon terminates path", input: "Iface[a/b;c.T]", want: "Iface[b;c.T]"},
		// Question mark is excluded
		{desc: "question mark terminates path", input: "Iface[a/b?c.T]", want: "Iface[b?c.T]"},
		// Colon is excluded
		{desc: "colon terminates path", input: "Iface[a/b:c.T]", want: "Iface[b:c.T]"},
		// Space terminates path (spec says "graphic chars without spaces")
		{desc: "space terminates path", input: "Iface[a/b c.T]", want: "Iface[b c.T]"},
		// Unicode replacement character U+FFFD is excluded
		{desc: "replacement char terminates path", input: "Iface[a/b\uFFFDc.T]", want: "Iface[b\uFFFDc.T]"},

```

**Comment by josharian** [comment]
> why did you decide to have non-path characters terminate the path segment rather than just have them be invalid?

Time box expired, decided to just ship, same as above.

File: impl_test.go [hunk 7]
```diff
@@ -55,6 +56,17 @@ func TestFindInterface(t *testing.T) {
 		{input: "github.com/josharian/impl/testdata.GenericInterface1[*github.com/josharian/impl/testdata.Struct5]", path: "github.com/josharian/impl/testdata", typ: Type{Name: "GenericInterface1", Params: []string{"*testdata.Struct5"}}},
 		// Hyphenated package path (hyphens in path segments, not in final package name)
 		{input: "github.com/go-chi/chi.Router[github.com/some-org/pkg.SomeType]", path: "github.com/go-chi/chi", typ: Type{Name: "Router", Params: []string{"pkg.SomeType"}}},
+
+		// Quoted path edge cases - unbalanced/double quotes
+		//// missing closing quote
```
Comment (josharian) [inline impl_test.go:61]: Add a comment above these explaining why they are here-but-commented-out.

File: impl.go [hunk 8]
```diff
@@ -80,38 +80,123 @@ func parseType(in string) (Type, error) {
...
-		if n == -1 {
+		if !more {
 			break
 		}
-		runes = runes[n:]
 	}
-	return string(out)
+
+	// We want balanced quotes for our paths
+	if quotesRemoved % 2 != 0 {
```
Comment (josharian) [inline impl.go:129]: nice

File: impl.go [hunk 9]
```diff
@@ -80,38 +80,123 @@ func parseType(in string) (Type, error) {
...
+		return "", fmt.Errorf("unbalanced quotes")
+	}
+
+	return string(out), nil
+}
+
+func getNonPathSeg(runes []rune, quotesRemoved *int) (seg []rune, remain []rune, more bool) {
```
Comment (josharian) [inline impl.go:136]: please return quotesRemoved instead of passing pointer. and it looks like it can be a bool (singular quoteRemoved), not an int, which makes it clearer.

File: impl.go [hunk 10]
```diff
@@ -80,38 +80,123 @@ func parseType(in string) (Type, error) {
...
+	}
+	// Trim a quote from the segment
+	lenPreTrim := len(seg)
+	seg = trimPathSeg(seg)
+	// If a quote was removed, increment the number of quotes removed
+	// This is for checking that the quotations are balanced
+	*quotesRemoved += lenPreTrim - len(seg)
+	// If there are no path like characters, we are done
+	if n == -1 {
+		return seg, []rune{}, false
```
Comment (josharian) [inline impl.go:152]: use nil instead of []rune{}

File: impl.go [hunk 11]
```diff
@@ -80,38 +80,123 @@ func parseType(in string) (Type, error) {
...
+	}
+	remain = runes[n:]
+	return seg, remain, true
+}
+
+func getPathSeg(runes []rune) (seg []rune, remain []rune, more bool) {
```
Comment (josharian) [inline impl.go:158]: `seg, remain []rune`

File: impl.go [hunk 12]
```diff
@@ -80,38 +80,123 @@ func parseType(in string) (Type, error) {
...
+	}
+	// if there is no non-path like characters, we are done
+	if n == -1 {
+		return seg, []rune{}, false
+	}
+	remain = runes[n:]
+	return seg, remain, true
+}
+
+func checkForQuote(p []rune) bool {
```
Comment (josharian) [inline impl.go:179]: call this hasQuote

File: impl.go [hunk 13]
```diff
@@ -80,38 +80,123 @@ func parseType(in string) (Type, error) {
...
+	if len(p) == 0 {
+		return false
+	}
+	return p[0] == '"' || p[len(p)-1] == '"'
+}
+
+func trimPathSeg(p []rune) []rune {
```
Comment (josharian) [inline impl.go:186]: trimQuote

File: impl.go [hunk 14]
```diff
@@ -80,38 +80,123 @@ func parseType(in string) (Type, error) {
...
+//   c. Strip second quote
+// 3. etc
+func stripPaths(in string) (string, error) {
+	remain := []rune(in)
+	out := make([]rune, 0, len(remain))
+	quotesRemoved := 0
+
+	for len(remain) > 0 {
+		var seg []rune
+		var more bool
```
Comment (josharian) [inline impl.go:108]: very optional: change more to done, because done is marginally nicer than !more as a condition.

**Review by josharian — COMMENTED** [review]
OK, basically there. bunch of little nits. if you want, say the word and i'll make the last few changes myself. otherwise, comments below. things that occur multiple places i've noted only once. oh, and please run gofumpt at the end.

**Comment by EliSauder** [comment]
@josharian I realized, without handling those test cases for findInterfaces, this won't actually resolve #13. I don't think it makes sense to merge in without that. So, I'll take a look at it along with your comments.
